external-crates/move/documentation/book/src/control-flow/pattern-matching.md

```diff
@@ -0,0 +1,618 @@
+not exhaustive, the compiler will raise an error. Note that any arm with a guard expression does not
+contribute to match exhaustion, as it may fail to match at runtime.
+
+As an example, if we were to match on a `u8` then in order for the match to be exhaustive we would
+need to match on _every_ number from 0 to 255 inclusive, or a wildcard or variable pattern would
+need to be present. Similarly if we were to match on a `bool` then we would need to match on both
+`true` and `false`, or a wildcard or variable pattern would need to be present.
+
+For structs, since there is only one type of constructor for the type, only one constructor needs to
+be matched, but the fields within the struct need to be matched exhaustively as well. Conversely,
+enums may define multiple variants, and each variant must be matched (including any sub-fields) in
+order for the match to be considered exhaustive.
+
+Since underscores and variables match anything, they count as matching all values of the type they
+are matching on in that position. Additionally, the multi-arity wildcard pattern `..` can be used to
+match on multiple values within a struct or enum variant.
+
+To see some examples of _non-exhaustive_ matches, consider the following:
+
+```move
+public enum MyEnum {
+    Variant(u64, bool),
+    OtherVariant(bool, u64),
+}
+
+public struct Pair<T>(T, T)
+
+fun f(x: MyEnum): u8 {
+    match (x) {
+        MyEnum::Variant(1, true) => 1,
+        MyEnum::Variant(_, _) => 1,
+        MyEnum::OtherVariant(_, 3) => 2,
+        // ERROR: not exhaustive as the value `MyEnum::OtherVariant(_, 4)` is not matched.
+    }
+}
+
+fun match_pair_bool(x: Pair<bool>): u8 {
+    match (x) {
+        Pair(true, true) => 1,
+        Pair(true, false) => 1,
+        Pair(false, false) => 1,
+        // ERROR: not exhaustive as the value `Pair(false, true)` is not matched.
+    }
+}
+```
+
+These examples can then be made exhaustive by adding a wildcard pattern to the end of the match arm,
+or by fully matching on the remaining values:
+
+```move
+fun f(x: MyEnum): u8 {
+    match (x) {
+        MyEnum::Variant(1, true) => 1,
+        MyEnum::Variant(_, _) => 1,
+        MyEnum::OtherVariant(_, 3) => 2,
+        // Now exhaustive since this will match all values of MyEnum::OtherVariant
+        MyEnum::OtherVariant(..) => 2,
+
+    }
+}
+
+fun match_pair_bool(x: Pair<bool>): u8 {
+    match (x) {
+        Pair(true, true) => 1,
+        Pair(true, false) => 1,
+        Pair(false, false) => 1,
+        // Now exhaustive since this will match all values of Pair<bool>
+        Pair(false, true) => 1,
+    }
+}
+```
+
+## Guards
+
+As mentioned above you can add a guard to a match arm by adding an `if` clause after the pattern.
+This guard will run _after_ the pattern has been matched but _before_ the expression on the
+right-hand-side of the arrow is evaluated. If the guard expression evaluates to `true` then the
```

Comment:
```suggestion
right hand side of the arrow is evaluated. If the guard expression evaluates to `true` then the
```
I don't think they're neede, but if you really dig dashes, use "right-hand side" instead.